src/Microsoft.ML.Transforms/Text/WordEmbeddingsTransform.cs

```diff
@@ -0,0 +1,446 @@
+                    ensureModel.Wait();
+                    var errorResult = ResourceManagerUtils.GetErrorMessage(out var errorMessage, ensureModel.Result);
+                    if (errorResult != null)
+                    {
+                        var directory = Path.GetDirectoryName(errorResult.FileName);
+                        var name = Path.GetFileName(errorResult.FileName);
+                        throw ch.Except($"{errorMessage}\nModel file for Word Embedding transform could not be found! " +
+                            $@"Please copy the model file '{name}' from '{url}' to '{directory}'.");
+                    }
+                    return ensureModel.Result.FileName;
+                }
+            }
+            throw Host.Except($"Can't map model kind = {kind} to specific file, please refer to https://aka.ms/MLNetIssue for assistance");
+        }
+
+        private Model GetVocabularyDictionary(out int dimension)
+        {
+            dimension = 0;
+            if (!File.Exists(_modelFileNameWithPath))
+                throw Host.Except("Custom word embedding model file '{0}' could not be found for Word Embeddings transform.", _modelFileNameWithPath);
+
+            if (_vocab.ContainsKey(_modelFileNameWithPath) && _vocab[_modelFileNameWithPath] != null)
+            {
+                if (_vocab[_modelFileNameWithPath].TryGetTarget(out Model model))
+                {
+                    dimension = model.Dimension;
+                    return model;
+                }
+            }
+
+            lock (_embeddingsLock)
+            {
+                if (_vocab.ContainsKey(_modelFileNameWithPath) && _vocab[_modelFileNameWithPath] != null)
+                {
+                    if (_vocab[_modelFileNameWithPath].TryGetTarget(out Model modelObject))
+                    {
+                        dimension = modelObject.Dimension;
+                        return modelObject;
+                    }
+                }
+
+                Model model = null;
+                using (StreamReader sr = File.OpenText(_modelFileNameWithPath))
+                {
+                    string line;
+                    int lineNumber = 1;
+                    char[] delimiters = { ' ', '\t' };
+                    using (var ch = Host.Start(LoaderSignature))
+                    using (var pch = Host.StartProgressChannel("Building Vocabulary from Model File for Word Embeddings Transform"))
+                    {
+                        var header = new ProgressHeader(new[] { "lines" });
+                        pch.SetHeader(header, e => e.SetProgress(0, lineNumber));
+                        string firstLine = sr.ReadLine();
+                        while ((line = sr.ReadLine()) != null)
+                        {
+                            if (lineNumber >= _linesToSkip)
+                            {
+                                string[] words = line.TrimEnd().Split(delimiters);
```

Comment:
Reading/parsing the model file is slow. Do you happen to see any places for perf improvement? Perf isn't intended to be part of this PR, but if there's something obvious...